trail_detection_node/trail_detection_node/v2_trailDetectionNode.py

```diff
@@ -12,241 +12,635 @@
+                Note: first col indicates pixel along width of image, and second col indicates along height.
+
+        Returns:
+            numpy.ndarray: UV coordinates of the route.
+        """
+
+        # Use camera properties to convert pixels to uv
+        uv_route = np.empty_like(route_indices).astype(float) #N, 2 
+        uv_route[:, 0] = (route_indices[:, 0] - self.c_x) / self.f_x
+        uv_route[:, 1] = (route_indices[:, 1] - self.c_y) / self.f_y
+        return uv_route
+    
+    def compute_route_uv_and_img(self, camera_msg: Image) -> tuple[np.ndarray, np.ndarray]:
+        """
+        Computes the planned route in uv coordinates and produces the image for visualization.
+
+        Args:
+            camera_msg (sensor_msgs.msg.Image): Camera message.
+
+        Returns:
+            tuple: Tuple containing route in uv coordinates(numpy.ndarray) and image with trail and centreline (numpy.ndarray).
+        """
+
+        def get_rgb_undistorted_img(camera_msg: Image) -> np.ndarray:
+            """
+            Processes the camera message to retrieve the undistorted RGB image.
+
+            Args:
+                camera_msg (sensor_msgs.msg.Image): Camera message.
+
+            Returns:
+                numpy.ndarray: Undistorted RGB image.
+            """
+            # process camera msg and set to RGB
+            cv_image = self.bridge.imgmsg_to_cv2(camera_msg, desired_encoding='passthrough')
+            cv_image = cv2.cvtColor(cv_image, cv2.COLOR_BGR2RGB)
+
+            # Undistort image
+            if not self.image_height:
+                self.image_height, self.image_width = cv_image.shape[:2]
+            new_K, _ = cv2.getOptimalNewCameraMatrix(CAM_INTRINSIC_K, RADIAL_DISTORTION, (self.image_width, self.image_height), 1, (self.image_width, self.image_height))  
+            cv_image = cv2.undistort(cv_image, CAM_INTRINSIC_K, RADIAL_DISTORTION, None, new_K)
+            return cv_image
+
+        def pre_process_img(image: np.ndarray, pre_proc_blur_k_size: int, brightness: int) -> np.ndarray:
+            """
+            Pre-processes the image for segmentation model.
+            Applies equalization, Gaussian Blur, and brightness increase.
+
+            Args:
+                image (numpy.ndarray): RGB undistorted input image.
+                pre_proc_blur_k_size (int): Kernel size for Gaussian blur.
+                brightness (int): Brightness adjustment value.
+
+            Returns:
+                numpy.ndarray: Pre-processed image.
+            """
+
+            # Split the image into its color channels
+            r, g, b = cv2.split(image)
+
+            # Equalize the histograms for each channel
+            r_eq = cv2.equalizeHist(r)
+            g_eq = cv2.equalizeHist(g)
+            b_eq = cv2.equalizeHist(b)
+
+            # Merge the channels
+            image = cv2.merge((r_eq, g_eq, b_eq))
+
+            # Apply Gaussian Blur
```

Comment:
The images are blurred so that some small features in the image are less distinguishable and cause less issues for the segmentation, such as small leaves, cracks in the sidewalk, and any sharp changes in colour/shade in the sidewalk. I had not considered median filter, should I move to that approach instead of Gaussian blur?

Reply:
With the old camera, we got artefacts in low-light images where you would get pixels with random colours. In that case, median filtering might help as it removes the outlier values and tends to be better at preserving edges. But if the issue is excessive edges / small details and you want to remove those, Gaussian filtering is a good solution.